src/App.tsx

```diff
@@ -1,250 +1,219 @@
+                    setError(err);
+                    setForceParsingError(true);
+                }
+            }
+        })();
+        /* Clean-up function called if there is a new file state while parsing is still in progress.
+         * Effectively abandons stale results. */
+        return () => {fileHasNotChanged = false};
+    }, [forceFile]);
+
+    // --------------------------------------- Animation & Animation Control Data --------------------------------------
+
+    /* Start, current, and end frames from data */
+    const frameStart = 0;
+    const [frame, setFrame] = useState(0);
+    const [frameEnd, setEnd] = useState(0);
+
+    /* User-selected start and end frames (for cropping) */
+    const [frameCropStart, setCropStart] = useState(0);
+    const [frameCropEnd, setCropEnd] = useState(0);
+
+    /* Current frame reference for use in dependency arrays where we don't want to trigger on every frame change */
+    const frameRef = useStateRef(frame);
+
+    /* Flags for playback */
+    const [playing, setPlaying] = useState(false);
+    const [loopPlayback, setLoopPlayback] = useState(true);
+
+    /* Handle on most recent animation loop, so it may be cancelled */
+    const animationRef = useRef<number>();
+
+    /* The time for one marker/force frame in ms */
+    const timeStepMillis = useMemo(() => {
+        if (markerFileData.frames.length<2) return null;
+        return markerFileData.frames[1].time * MILLIS_PER_SEC; //convert from secs to millis
+    }, [markerFileData]);
+
+    /* Time of previous animation loop iteration (most recent render).
+     * For finding elapsed time between loops to determine which marker/force frame to render next. */
+    const lastRepaintTimeRef = useRef<DOMHighResTimeStamp|null>(null);
+
+    /* The cumulative elapsed time that has not yet been animated.
+     * Decremented by timeStepMillis for every marker frame increment. */
+    const interFrameTimeRef = useRef(0);
+
+    /* Tail recursive loop, called once per next available browser repaint (at rate of user's display's refresh rate),
+     * that handles updating the marker/force frame number (to drive the animation) by the correct amount according to
+     * the time elapsed between repaints. This allows the animation to play at 1x speed. */
+    const animationLoop = useCallback((currentTime: DOMHighResTimeStamp) => {
+        /* If file has not been parsed for the time step, there isn't enough info to know how to animate */
+        if (!timeStepMillis) return;
+        /* When paused, pretend no time has elapsed */
+        const elapsedTime = (lastRepaintTimeRef.current!==null) ? currentTime-lastRepaintTimeRef.current : 0;
+        /* When playing, before next repaint, increment frame by the correct amount for the elapsed time since last paint */
+        for (interFrameTimeRef.current += elapsedTime; interFrameTimeRef.current > timeStepMillis; interFrameTimeRef.current -= timeStepMillis) {
```

Comment:
this is a confusing way to write this, I would at least move out the addition

Reply:
I admit the readability here probably comes down to personal preference. I'm hoping the others can weigh in on what they find easiest to follow. My reasoning is that this format clearly separates loop control variables from the work happening in the body of the loop, so once the reader digests the control flow of a single loop, they only need to focus on the sole effect of the frame being incremented.

I'll put the original loop here so the others can easily compare it without digging through the diff:
```typescript
const animationLoop = useCallback(() => {
	if(timeStep !== null) {
		const currentTime = performance.now();
		if(lastTimeRef.current !== null) {
			const elapsedTime = currentTime - lastTimeRef.current;
			interFrameTimeRef.current += elapsedTime;
			while(interFrameTimeRef.current > timeStep) {
				interFrameTimeRef.current -= timeStep;
				setFrame(current => {
					if(current + 1 < markerFileData.frames.length && current + 1 < frameEnd) return current + 1;
					else return frameStart;
				});
			}
		}
		lastTimeRef.current = currentTime;
	}

	animationRef.current = requestAnimationFrame(animationLoop);
}, [markerFileData, timeStep, frameEnd, frameStart]);
```
And the refactored version:
```typescript
/* Tail recursive loop, called once per next available browser repaint (at rate of user's display's refresh rate),
 * that handles updating the marker/force frame number (to drive the animation) by the correct amount according to
 * the time elapsed between repaints. This allows the animation to play at 1x speed. */
const animationLoop = useCallback((currentTime: DOMHighResTimeStamp) => {
    /* If file has not been parsed for the time step, there isn't enough info to know how to animate */
    if (!timeStepMillis) return;
    /* When paused, pretend no time has elapsed */
    const elapsedTime = (lastRepaintTimeRef.current!==null) ? currentTime-lastRepaintTimeRef.current : 0;
    /* When playing, before next repaint, increment frame by the correct amount for the elapsed time since last paint */
    for (interFrameTimeRef.current += elapsedTime; interFrameTimeRef.current > timeStepMillis; interFrameTimeRef.current -= timeStepMillis) {
        setFrame(current => {
            if (current+1<=frameCropEnd) return current+1; //next frame is in range, proceed
            else if (loopPlayback) return frameCropStart; //not in range: wrap around to start if looping
            else {setPlaying(false); return current;} //not in range: stop playing on final frame if not looping
        });
    }
    /* While playing, loop forever with tail recursion */
    if (playing) {
        lastRepaintTimeRef.current = currentTime; //store time for getting elapsed time in next loop
        animationRef.current = requestAnimationFrame(animationLoop); //get handle of new loop
    }
}, [playing, loopPlayback, timeStepMillis, frameCropStart, frameCropEnd]);
```